pulsar-client-cpp/lib/ClientImpl.cc

```diff
@@ -414,36 +402,32 @@ void ClientImpl::handleConsumerCreated(Result result, ConsumerImplBaseWeakPtr co
 
 Future<Result, ClientConnectionWeakPtr> ClientImpl::getConnection(const std::string& topic) {
     Promise<Result, ClientConnectionWeakPtr> promise;
-    lookupServicePtr_->lookupAsync(topic).addListener(std::bind(&ClientImpl::handleLookup, shared_from_this(),
-                                                                std::placeholders::_1, std::placeholders::_2,
-                                                                promise));
-    return promise.getFuture();
-}
 
-void ClientImpl::handleLookup(Result result, LookupDataResultPtr data,
-                              Promise<Result, ClientConnectionWeakPtr> promise) {
-    if (data) {
-        const std::string& logicalAddress =
-            clientConfiguration_.isUseTls() ? data->getBrokerUrlTls() : data->getBrokerUrl();
-        LOG_DEBUG("Getting connection to broker: " << logicalAddress);
-        const std::string& physicalAddress =
-            data->shouldProxyThroughServiceUrl() ? serviceUrl_ : logicalAddress;
-        Future<Result, ClientConnectionWeakPtr> future =
-            pool_.getConnectionAsync(logicalAddress, physicalAddress);
-        future.addListener(std::bind(&ClientImpl::handleNewConnection, shared_from_this(),
-                                     std::placeholders::_1, std::placeholders::_2, promise));
-    } else {
-        promise.setFailed(result);
+    const auto topicNamePtr = TopicName::get(topic);
+    if (!topicNamePtr) {
+        LOG_ERROR("Unable to parse topic - " << topic);
+        promise.setFailed(ResultInvalidTopicName);
+        return promise.getFuture();
     }
-}
 
-void ClientImpl::handleNewConnection(Result result, const ClientConnectionWeakPtr& conn,
-                                     Promise<Result, ClientConnectionWeakPtr> promise) {
-    if (result == ResultOk) {
-        promise.setValue(conn);
-    } else {
-        promise.setFailed(ResultConnectError);
-    }
+    auto self = shared_from_this();
+    lookupServicePtr_->getBroker(*topicNamePtr)
```

Comment:
We should pass a `std::shared_ptr` here, otherwise the object that `TopicNamePtr` points to might be invalid.

Reply:
But it should be okay in this PR. Because only the returned `std::string` is used in `getBroker` method. But if we applied the retry logic in future, it would be a problem.